service/gcs/bridge/bridge.go

```diff
@@ -14,385 +16,563 @@ import (
+		panic("bridge: nil handler func")
 	}
+
+	mux.Handle(id, HandlerFunc(handler))
 }
 
-func (b *bridge) loop() error {
-	var err error
-	b.commandConn, err = b.createAndConnectCommandConn()
-	if err != nil {
-		return err
+// Handler returns the handler to use for the given request type.
+func (mux *Mux) Handler(r *Request) Handler {
+	mux.mu.Lock()
+	defer mux.mu.Unlock()
+
+	if r == nil {
+		panic("bridge: nil request to handler")
 	}
-	defer b.commandConn.Close()
 
-	for {
-		// Messages from the HCS come as a header, which contains information
-		// about the message (including its size), and a message body.
-		// The header and body contain the operation to be performed, as well
-		// as any information needed to perform the operation.
-		message, header, err := readMessage(b.commandConn)
-		if err != nil {
-			return errors.Wrap(err, "failed to read a message from the HCS")
-		}
+	if mux.m == nil {
+		return NotSupportedHandler()
+	}
 
-		// Carry out the operations requested by the HCS.
-		// Each operation has its own helper function, each of which returns a
-		// response object.
-		var response interface{}
-		switch header.Type {
-		case prot.ComputeSystemCreateV1:
-			logrus.Info("received from HCS: ComputeSystemCreateV1")
-			response, err = b.createContainer(message)
-			if err != nil {
-				logrus.Error(err)
-			}
-		case prot.ComputeSystemExecuteProcessV1:
-			logrus.Info("received from HCS: ComputeSystemExecuteProcessV1")
-			response, err = b.execProcess(message)
-			if err != nil {
-				logrus.Error(err)
-			}
-		case prot.ComputeSystemShutdownForcedV1:
-			logrus.Info("received from HCS: ComputeSystemShutdownForcedV1")
-			response, err = b.killContainer(message)
-			if err != nil {
-				logrus.Error(err)
-			}
-		case prot.ComputeSystemShutdownGracefulV1:
-			logrus.Info("received from HCS: ComputeSystemShutdownGracefulV1")
-			response, err = b.shutdownContainer(message)
-			if err != nil {
-				logrus.Error(err)
-			}
-		case prot.ComputeSystemSignalProcessV1:
-			logrus.Info("received from HCS: ComputeSystemSignalProcessV1")
-			response, err = b.signalProcess(message)
-			if err != nil {
-				logrus.Error(err)
-			}
-		case prot.ComputeSystemGetPropertiesV1:
-			logrus.Info("received from HCS: ComputeSystemGetPropertiesV1")
-			response, err = b.listProcesses(message)
-			if err != nil {
-				logrus.Error(err)
+	var h Handler
+	var ok bool
+	if h, ok = mux.m[r.Header.Type]; !ok {
+		return NotSupportedHandler()
+	}
+
+	return h
+}
+
+// ServeMsg dispatches the request to the handler whose
+// type matches the request type
+func (mux *Mux) ServeMsg(w ResponseWriter, r *Request) {
+	h := mux.Handler(r)
+	h.ServeMsg(w, r)
+}
+
+// Request is the bridge request that has been sent
+type Request struct {
+	Header  *prot.MessageHeader
+	Message []byte
+}
+
+// ResponseWriter is the dispatcher used to construct the Bridge response
+type ResponseWriter interface {
+	// Header is the request header that was requested
```

Comment:
The go linter threw me off here. Your not supposed to use punctuation in errors but in comments it doesn't care. I added them.